addons/mail/tests/discuss/test_discuss_mail_presence.py

```diff
@@ -49,52 +49,40 @@ def _receive_presence(self, sender, recipient):
-        guest = self.env["mail.guest"].create({"name": "Guest"})
-        bob = new_test_user(self.env, login="bob_user", groups="base.group_user")
-        # Guest should not receive users's presence: no common channel.
-        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
-            self._receive_presence(sender=bob, recipient=guest)
-        channel = self.env["discuss.channel"]._create_channel(group_id=None, name="General")
-        channel._add_members(guests=guest, users=bob)
-        # Now that they share a channel, guest should receive users's presence.
-        self._receive_presence(sender=bob, recipient=guest)
-
-        other_guest = self.env["mail.guest"].create({"name": "OtherGuest"})
-        # Guest should not receive guest's presence: no common channel.
-        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
-            self._receive_presence(sender=other_guest, recipient=guest)
-        channel._add_members(guests=other_guest)
-        # Now that they share a channel, guest should receive guest's presence.
-        self._receive_presence(sender=other_guest, recipient=guest)
-        self.assertEqual(other_guest.im_status, "online")
-
-    def test_receive_presences_as_portal(self):
-        portal = new_test_user(self.env, login="portal_user", groups="base.group_portal")
-        bob = new_test_user(self.env, login="bob_user", groups="base.group_user")
-        # Portal should not receive users's presence: no common channel.
-        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
-            self._receive_presence(sender=bob, recipient=portal)
-        channel = self.env["discuss.channel"]._create_channel(group_id=None, name="General")
-        channel._add_members(users=portal | bob)
-        # Now that they share a channel, portal should receive users's presence.
-        self._receive_presence(sender=bob, recipient=portal)
-
+    def test_presence_access(self):
+        internal_user = new_test_user(self.env, login="internal_user", groups="base.group_user")
+        other_internal_user = new_test_user(
+            self.env, login="other_internal_user", groups="base.group_user"
+        )
+        portal_user = new_test_user(self.env, login="portal_user", groups="base.group_portal")
         guest = self.env["mail.guest"].create({"name": "Guest"})
-        # Portal should not receive guest's presence: no common channel.
-        with self.assertRaises(ws._exceptions.WebSocketTimeoutException):
-            self._receive_presence(sender=guest, recipient=portal)
-        channel._add_members(guests=guest)
-        # Now that they share a channel, portal should receive guest's presence.
-        self._receive_presence(sender=guest, recipient=portal)
-        self.assertEqual(guest.im_status, "online")
+        test_cases = [
+            # from - to - has_token - allowed
+            (internal_user, portal_user, False, False),
```

Comment:
Isn't first 2 columns inverted? From internal user should always be True

Reply:
Comments might be misleading: receiving presence from internal_user to the portal_user should only be allowed if the token is passed.

Reply:
I would write the test differently, the first column should be the current user attempting the read, and the 2nd should be the target user.